File: src/com/morph/engine/physics/components/Transform3D.java
package com.morph.engine.physics.components;

import com.morph.engine.math.Matrix4f;
import com.morph.engine.math.Quaternion;
import com.morph.engine.math.Vector3f;

public class Transform3D extends Transform {
	private Vector3f position;
	private Quaternion orientation;
	private Vector3f scale;
	
	public Transform3D(Vector3f position, Quaternion orientation, Vector3f scale) {
		this.position = position;
		this.orientation = orientation;
		this.scale = scale;
	}
	
	public Transform3D(Vector3f position, Quaternion orientation) {
		this(position, orientation, new Vector3f(1, 1, 1));
	}
	
	public Transform3D(Vector3f position) {
		this(position, new Quaternion(new Vector3f(0, 1, 0), 0), new Vector3f(1, 1, 1));
	}
	
	public Transform3D(Vector3f position, Vector3f scale) {
		this(position, new Quaternion(new Vector3f(0, 1, 0), 0), scale);
	}

	@Override
	public Matrix4f getTranslationMatrix() {
		return position.asTranslationMatrix();
	}

	@Override
	public Matrix4f getRotationMatrix() {
		return orientation.asRotationMatrix();
	}

	@Override
	public Matrix4f getScaleMatrix() {
		return scale.asScaleMatrix();
	}

	@Override
	public void translate(Matrix4f translation) {
		position = position.add(translation.asTranslationVector());
	}

	@Override
	public void rotate(Quaternion rotation) {
		orientation = orientation.mul(rotation);
	}

	@Override
	public void scale(Matrix4f scaling) {
		scale = scale.add(scaling.asScaleVector());
	}
	
	public void translate(Vector3f translation) {
		position = position.add(translation);
	}

	public void scale(Vector3f scaling) {
		scale = scale.add(scaling);
	}

	public Vector3f getPosition() {
		return position;
	}

	public void setPosition(Vector3f position) {
		this.position = position;
	}

	public Quaternion getOrientation() {
		return orientation;
	}

	public void setOrientation(Quaternion orientation) {
		this.orientation = orientation;
	}

	public Vector3f getScale() {
		return scale;
	}

	public void setScale(Vector3f scale) {
		this.scale = scale;
	}
	
	public Transform3D clone() {
		return new Transform3D(position, orientation, scale);
	}
}

File: src/main/kotlin/com/morph/engine/physics/components/Transform3D.kt
package com.morph.engine.physics.components

import com.morph.engine.math.Matrix4f
import com.morph.engine.math.Quaternion
import com.morph.engine.math.Vector3f

data class Transform3D @JvmOverloads constructor(var position: Vector3f = Vector3f(0f, 0f ,0f), var orientation: Quaternion = Quaternion(Vector3f(0f, 1f, 0f), 0f), var scale: Vector3f = Vector3f(1f, 1f, 1f)) : Transform() {
    override val translationMatrix: Matrix4f
        get() = position.asTranslationMatrix()

    override val rotationMatrix: Matrix4f
        get() = orientation.asRotationMatrix()

    override val scaleMatrix: Matrix4f
        get() = scale.asScaleMatrix()

    override fun translate(translation: Matrix4f) {
        position += translation.asTranslationVector()
    }

    override fun rotate(rotation: Quaternion) {
        orientation *= rotation
    }

    override fun scale(scale: Matrix4f) {
        this.scale += scale.asScaleVector()
    }

    fun translate(translation: Vector3f) {
        position += translation
    }

    fun scale(scaling: Vector3f) {
        scale += scaling
    }
}
